scalding-serialization/src/main/scala/com/twitter/scalding/serialization/ComplexHelper.scala

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.twitter.scalding.serialization
+
+import com.twitter.scalding.serialization.JavaStreamEnrichments._
+
+import java.io.{ ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream }
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+abstract class ComplexHelper[T] extends HasUnsafeCompareBinary[T] {
+  def staticSize: Option[Int] = None
+
+  protected def dynamicSizeWithoutLen(e: T): Option[Int]
+  final def dynamicSize(e: T) =
+    if (staticSize.isDefined) staticSize
+    else
+      dynamicSizeWithoutLen(e).map { e =>
+        e + posVarIntSize(e)
+      }
+  final def unsafeSize(t: T): Option[Int] = dynamicSizeWithoutLen(t)
+
+  /**
+   * This is the worst case: we have to serialize in a side buffer
+   * and then see how large it actually is. This happens for cases, like
+   * string, where the cost to see the serialized size is not cheaper than
+   * directly serializing.
+   */
+  private[this] def noLengthWrite(element: T, outerOutputStream: OutputStream): Unit = {
+    // Start with pretty big buffers because reallocation will be expensive
+    val baos = new ByteArrayOutputStream(512)
+    unsafeWrite(baos, element)
```

Comment:
this implies that `unsafeWrite` means no size. Can we add that to the comments below?

Reply:
It implies no outer size, it can be a bit wasteful. I'm adding a comment to the definition of this method that I hope will help here a little:
```
  // This will write out the interior data as a blob with no prepended length
  // This means binary compare cannot skip on this data.
  // However the contract remains that one should be able to _read_ the data
  // back out again.
  def unsafeWrite(out: java.io.OutputStream, t: T): Unit
```